src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,178 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private int size = 0;
+    private Node head;
+    private Node tail;
+
     @Override
     public void add(T value) {
+        if (size == 0) {
+            head = new Node(null, value, null);
+            tail = head;
+        } else if (size > 0) {
+            Node lastNode = tail;
+            tail = new Node(lastNode, value, null);
+            lastNode.next = tail;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (size >= 0 && index == size) {
+            add(value);
+        } else {
+            isIndexExist(index);
+            Node currentNode = findNodeByIndex(index);
+            if (index == 0) {
+                head = new Node(currentNode.prev, value, currentNode);
+                size++;
+                return;
+            }
+            Node previousNod = currentNode.prev;
+            currentNode.prev = new Node(currentNode.prev, value, currentNode);
+            previousNod.next = currentNode.prev;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        isIndexExist(index);
+        Node node = findNodeByIndex(index);
+        return (T) node.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        isIndexExist(index);
+        Node node = findNodeByIndex(index);
+        T oldValue = (T) node.value;
+        node.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        isIndexExist(index);
+        T value = (T) findNodeByIndex(index).value;
+        unlink(findNodeByIndex(index));
+        return value;
     }
 
     @Override
     public boolean remove(T object) {
+        Node node = head;
+        for (int i = 0; i < size; i++) {
+            if (needToRemove(object, node)) {
+                unlink(node);
+                return true;
+            }
+            node = node.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private static class Node<T> {
+        private T value;
+        private Node<T> next;
+        private Node<T> prev;
+
+        private Node(Node<T> prev, T element, Node<T> next) {
+            this.value = element;
+            this.next = next;
+            this.prev = prev;
+        }
+    }
+
+    private void isIndexExist(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException();
+        }
+    }
+
+    private boolean needToRemove(T object, Node node) {
```

Comment:
let's change this method a bit to make it more pretty and understandable:
* since this method basically compares values, would be nice receive values of the same type
* if all it does is comparing, then let's call it accordingly
```suggestion
    private boolean areValuesEqual(T first, T second) {
```